kotlin/codegen/src/test/java/com/squareup/moshi/kotlin/codegen/ksp/JsonClassSymbolProcessorTest.kt

```diff
@@ -0,0 +1,824 @@
+        "source.kt",
+        """
+          package test
+          import com.squareup.moshi.JsonClass
+          import com.squareup.moshi.JsonQualifier
+          import kotlin.annotation.AnnotationRetention.RUNTIME
+          import kotlin.annotation.AnnotationTarget.PROPERTY
+          import kotlin.annotation.Retention
+          import kotlin.annotation.Target
+
+          @Retention(RUNTIME)
+          @Target(PROPERTY)
+          @JsonQualifier
+          annotation class UpperCase
+
+          @JsonClass(generateAdapter = true)
+          class ClassWithQualifier(@UpperCase val a: Int)
+          """
+      )
+    )
+    assertThat(result.exitCode).isEqualTo(KotlinCompilation.ExitCode.COMPILATION_ERROR)
+    assertThat(result.messages).contains("JsonQualifier @UpperCase must support FIELD target")
+  }
+
+  @Test
+  fun nonRuntimeQualifier() {
+    val result = compile(
+      kotlin(
+        "source.kt",
+        """
+          package test
+          import com.squareup.moshi.JsonClass
+          import com.squareup.moshi.JsonQualifier
+          import kotlin.annotation.AnnotationRetention.BINARY
+          import kotlin.annotation.AnnotationTarget.FIELD
+          import kotlin.annotation.AnnotationTarget.PROPERTY
+          import kotlin.annotation.Retention
+          import kotlin.annotation.Target
+
+          @Retention(BINARY)
+          @Target(PROPERTY, FIELD)
+          @JsonQualifier
+          annotation class UpperCase
+
+          @JsonClass(generateAdapter = true)
+          class ClassWithQualifier(@UpperCase val a: Int)
+          """
+      )
+    )
+    assertThat(result.exitCode).isEqualTo(KotlinCompilation.ExitCode.COMPILATION_ERROR)
+    assertThat(result.messages).contains("JsonQualifier @UpperCase must have RUNTIME retention")
+  }
+
+  @Test
+  fun `TypeAliases with the same backing type should share the same adapter`() {
+    val result = compile(
+      kotlin(
+        "source.kt",
+        """
+          package test
+          import com.squareup.moshi.JsonClass
+
+          typealias FirstName = String
+          typealias LastName = String
+
+          @JsonClass(generateAdapter = true)
+          data class Person(val firstName: FirstName, val lastName: LastName, val hairColor: String)
+          """
+      )
+    )
+    assertThat(result.exitCode).isEqualTo(KotlinCompilation.ExitCode.OK)
+
+    // We're checking here that we only generate one `stringAdapter` that's used for both the
+    // regular string properties as well as the the aliased ones.
+    // TODO loading compiled classes from results not supported in KSP yet
+//    val adapterClass = result.classLoader.loadClass("PersonJsonAdapter").kotlin
+//    assertThat(adapterClass.declaredMemberProperties.map { it.returnType }).containsExactly(
+//      JsonReader.Options::class.createType(),
+//      JsonAdapter::class.parameterizedBy(String::class)
+//    )
+  }
+
+  @Test
+  fun `Processor should generate comprehensive proguard rules`() {
```

Comment:
This is not a good test, but it serves as nice raw material from which good tests could be created. Rather than covering an assortment of features in one scattershot test, we aught to test each of the interesting features each in their own tests.

Reply:
Let's do a dedicated PR for that, the existing apt tests are the same and it'd be pretty involved

Reply:
That'd be great! I'm more ranting than anything.